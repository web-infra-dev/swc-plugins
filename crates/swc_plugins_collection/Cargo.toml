[package]
name = "swc_plugins_collection"
version = "0.1.0"
edition = "2021"


[dependencies]

serde_json = { workspace = true }
serde = { workspace = true }
swc_core = { workspace = true, features = [
    "base",
    "common",
    "ecma_ast",
    "ecma_visit",
    "ecma_transforms",
    "ecma_parser",
] }
swc_plugins_utils = { path = "../swc_plugins_utils" }
swc_plugins_core = { path = "../swc_plugins_core" }

# plugins list 
modularize_imports = "0.54.0"
plugin_react_const_elements = { path = "../plugin_react_const_elements" }
swc_plugin_import = { path = "../plugin_import" }
plugin_lock_corejs_version = { path = "../plugin_lock_corejs_version" }
swc_plugin_lodash = { path = "../plugin_lodash" }
swc_plugin_react_utils = { path = "../plugin_react_utils" }
plugin_remove_es_module_mark = { path = "../plugin_remove_es_module_mark" }
plugin_ssr_loader_id = { path = "../plugin_ssr_loader_id" }
plugin_config_routes = { path = "../plugin_config_routes" }
styled_components = "0.81.0"
styled_jsx = "0.58.0"
swc_emotion = "0.57.0"
swc_plugin_loadable_components = { path = "../plugin_loadable_components" }

[dev-dependencies]
insta = "1.18.2"
