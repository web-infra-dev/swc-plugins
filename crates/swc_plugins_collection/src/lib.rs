#![feature(let_chains)]

pub mod pass;
pub mod types;
pub use modularize_imports;
pub use plugin_config_routes;
pub use plugin_lock_corejs_version;
pub use plugin_ssr_loader_id;
pub use styled_components as plugin_styled_components;
pub use styled_jsx as plugin_styled_jsx;
pub use swc_emotion as plugin_emotion;
pub use swc_plugin_import;
pub use swc_plugin_loadable_components;
pub use swc_plugin_lodash;
pub use swc_plugin_react_utils;
