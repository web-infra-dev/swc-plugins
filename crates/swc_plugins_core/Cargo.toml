[package]
edition = "2021"
name = "swc_plugins_core"
version = "0.1.0"

[features]
plugin = [
  "swc_core/plugin_transform_host_native_filesystem_cache",
  "swc_core/plugin_transform_host_native",
]

[dependencies]
anyhow = { workspace = true }
swc_core = { workspace = true, features = [
  "base",
  "common_concurrent",
  "css_ast",
  "css_codegen",
  "css_minifier",
  "css_parser",
  "ecma_ast",
  "ecma_visit",
  "ecma_transforms",
  "ecma_parser",
  "ecma_minifier_concurrent",
] }
swc_plugins_utils = { path = "../swc_plugins_utils" }
serde_json = { workspace = true }
serde = { workspace = true }
regex = "1.6.0"

[dev-dependencies]
insta = "1.18.2"
