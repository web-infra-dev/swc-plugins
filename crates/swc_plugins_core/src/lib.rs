#![feature(let_chains)]
mod minify;
mod pass;
mod transform;

pub mod types;
pub use minify::minify;
pub use transform::transform;

// plugins
pub use modularize_imports;
pub use swc_plugin_import;
pub use plugin_lock_corejs_version;
pub use swc_plugin_lodash;
pub use swc_plugin_react_utils;
pub use styled_components as plugin_styled_components;
pub use styled_jsx as plugin_styled_jsx;
pub use swc_emotion as plugin_emotion;
