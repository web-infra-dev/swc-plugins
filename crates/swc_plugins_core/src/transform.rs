use std::{path::PathBuf, sync::Arc};

use anyhow::Result;
use swc_core::{
  base::{
    config::{self, Options},
    try_with_handler, Compiler, HandlerOpts, TransformOutput,
  },
  common::{comments::SingleThreadedComments, errors::ColorConfig, FileName, Mark, GLOBALS},
  ecma::{
    ast::EsVersion,
    parser::{Syntax, TsConfig},
    visit::Fold,
    // TODO current version too low
    // transforms::module::common_js::Config
  },
};
use swc_plugins_utils::PluginContext;

use crate::TransformFn;

/// As we initialize plugins at each transform,
/// Some plugins need very heavy work on the first
/// time, and if we can cache it, we should get better
/// performance.
/// A `config_hash` is the unique key representing a
/// specific `TransformConfig`.
/// transform don't care how you create this hash.
///
/// If you call `transform` from `nodejs`, this config hash
/// is unique for each `binding::Compiler`.
#[allow(clippy::too_many_arguments)]
pub fn transform<'a, E, P1, P2>(
  compiler: Arc<Compiler>,
  swc_config: &Options,
  extensions_config: &'a E,
  filename: impl Into<String>,
  code: &str,
  input_source_map: Option<String>,
  config_hash: Option<String>,
  transform_before_pass: TransformFn<'a, E, P1>,
  transform_after_pass: TransformFn<'a, E, P2>,
) -> Result<TransformOutput>
where
  E: 'a,
  P1: Fold + 'a,
  P2: Fold + 'a,
{
  GLOBALS.set(&Default::default(), || {
    let cm = compiler.cm.clone();
    let filename: String = filename.into();

    try_with_handler(
      cm.clone(),
      HandlerOpts {
        color: ColorConfig::Never,
        skip_filename: false,
      },
      |handler| {
        compiler.run_transform(handler, true, || {
          let cm_filename = if filename.is_empty() {
            FileName::Anon
          } else {
            FileName::Real(PathBuf::from(filename.clone()))
          };

          let fm = cm.new_source_file(cm_filename, code.to_string());

          let mut swc_config = swc_config.clone();
          swc_config.config.input_source_map = input_source_map.map(config::InputSourceMap::Str);
          swc_config.filename = filename.clone();

          let top_level_mark = swc_config.top_level_mark.unwrap_or_else(Mark::new);
          let unresolved_mark = Mark::new();

          swc_config.top_level_mark = Some(top_level_mark);
          let comments = SingleThreadedComments::default();
          let is_module = swc_config
            .config
            .is_module
            .unwrap_or(config::IsModule::Unknown);

          let program = compiler.parse_js(
            fm.clone(),
            handler,
            swc_config.config.jsc.target.unwrap_or(EsVersion::Es2022),
            swc_config.config.jsc.syntax.unwrap_or_else(|| {
              Syntax::Typescript(TsConfig {
                tsx: true,
                decorators: true,
                ..Default::default()
              })
            }),
            is_module,
            Some(&comments),
          )?;

          // TODO comments can be pass to `process_js_with_custom_pass` in next swc version
          let plugin_context = Arc::new(PluginContext {
            cm,
            file: fm.clone(),
            top_level_mark,
            unresolved_mark,
            comments: comments.clone(),
            config_hash,
            filename,
            cwd: swc_config.cwd.clone(),
          });

          compiler.process_js_with_custom_pass(
            fm,
            Some(program),
            handler,
            &swc_config,
            comments,
            // TODO pass comments to internal pass in next swc versions
            |_| transform_before_pass(extensions_config, &swc_config, plugin_context.clone()),
            |_| transform_after_pass(extensions_config, &swc_config, plugin_context.clone()),
          )
        })
      },
    )
  })
}
