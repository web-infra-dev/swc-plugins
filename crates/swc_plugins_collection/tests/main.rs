use std::sync::Arc;

use swc_core::{
  base::{config::Options, Compiler},
  common::SourceMap,
};
use swc_plugins_collection::{pass, types};
use swc_plugins_core::transform;

#[test]
fn test() {
  let code = "const a = {};
    const b = { ...a }";
  let swc_config = Options {
    config: serde_json::from_str(
      r#"{
          "jsc": {
            "externalHelpers": true
          }
        }"#,
    )
    .unwrap(),
    ..Default::default()
  };

  let extensions_config = types::Extensions::default();
  let res = transform(
    Arc::new(Compiler::new(Arc::new(SourceMap::default()))),
    &swc_config,
    &extensions_config,
    "",
    code,
    None,
    Some("".into()),
    pass::internal_transform_before_pass,
    pass::internal_transform_after_pass,
  )
  .unwrap();
  insta::assert_snapshot!("plugin-import", res.code);
}
