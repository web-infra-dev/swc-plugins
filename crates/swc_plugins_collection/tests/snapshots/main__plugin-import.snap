---
source: crates/swc_plugins_collection/tests/main.rs
expression: res.code
---
var _object_spread = require("@swc/helpers/_/_object_spread");
var a = {};
var b = _object_spread._({}, a);

